### ISSUE TITLE
Allow replacement

### ISSUE THREAD

**Issue by sils1297** [coala-analyzer/coala#167]
Ok, for now coala provides almost everything thats needed for diagnostic things. It would be nice to now go for automatically solve annoying mini-changes. We need some replacement result. In our prototypes we had some LineReplacement result; @smfn3321 and me thought quite some time about this and IIRC we concluded that it would be best to have some DiffResult where we could store a multiline diff.

@smfn3321 can you shed some more light on this?

**Comment by sils1297** [comment]
Digged up our old merger: https://github.com/sils1297/coala/blob/new/coalib/internal/modifying/Merger.py
Don't know if this'll be useful.

**Comment by sils1297** [comment]
and heres our current mockup: https://github.com/sils1297/coala/blob/master/doc/requirements/output_example

**Comment by smfn3321** [comment]
Well, oftentimes one gets only so far with changes in only one line - which is the freedom that we have at this point per result. It would be easier and simpler for the user to group several *changes* into *tasks* to then handle only one *result* per *task*. 

In return these results must then carry the information for several changes and one possible solution for this case would be to introduce a *diff-result*. A diff results carries informatino in form of a diff that specifies all differencies between the file before and after all changes of a task have been applied.

A possible case in which such action would be appropriate is the refactoring/renaming of variables. A variable name might be changed in several places and it would be countrproductive and dangerous to ask for each change separately as this may introduce breackage if the answers are not consistent. A diff-result may either be applied as a whole or not at all and therefore preserves consistency.

A problem that might be introduced by diff results is that several diffs might be conflicting which each other. Since a lot of the programs flow depends on the results, the work necessary for the implementation of yet another kind of them might also need to be considered.

**Comment by sils1297** [comment]
It is also important to note that diff results are not necessarily per
file! E.g. a parameter renaming will probably affect other files too.

Thus we probably don't want to go for a diff result but rather a kind of
patch result which effectively represents a patch (same thing as a commit
but without git ;) )
Am 19.01.2015 19:33 schrieb "Fabian Neuschmidt" <notifications@github.com>:

**Comment by smfn3321** [comment]
Okay I did not think of that until now, but this might even kinda support the idea that I had about the design of this. Note that we already talked about how it would make sense to ask the user at the end to confirm all changes by showing a complete diff of all changes?

well, I think both aspects encourage to work on some kind of diff during the run of the programm which can be applied at the end. Anyway:

## Design Proposal for Diff-Class
### Background
A diff can usually only be applied on the file that it was created on. If this file is modified, the diff can generally only be applied by knowledge of the source. A common solution to this problem is to create a second diff from the modified file and the original one and then use something like a three-way-merge to merge both diffs into the original.

For a large file this might create a lot of overheat and since we want to keep track of original line numbers this might prove tedious in the implementation.

### Proposed Solution
I propose to create a diff-class that can be created by bears and then returned as a member of a patch-result. This diff-class contains changes for *one* file that are saved in connection with their line-number. Patch-results should be able to contain several diff-classes since i.e. refactoring changes might be applicable to several files.

The class will provide methods for creation that can be used by bears i.e. .delete_line(line_nr). I propose three kinds of changes that can be made:
* deletion of a number of lines lines (resulting in several entries (one for each line) that this line has been deleted)
* addition of a number of lines (resulting in one entry (in between the line before and the one after the change) that these lines were added )
* changing of a line (resulting in an entry for this line containing a difflib.diff() for the change)

The class will provide methods for the addition of another diff-classe returning a single, unified diff-class containing all changes. Since diff-classes save changes by line number this should be relatively easy and more performant than a 3-way-merge.

I propose the following behavior if two changes are to be applied to one line:
* deletion on both sides: line deleted, no conflict
* deletion and change: controversial, line deleted, maybe warning, (eventually conflict?)
* deletion and addition after: both applied, no conflict
* change on both sides: change-diffs will be 3-way-merged if possible, MIGHT CONFLICT 
* change and addition after: both applied, no conflict
* addition after on both sides: WILL CONFLICT

This behavior will allow multiple changes to a line. This might be usefull in a case where an indentation fix should be applied as well as the renaming of a variable. Furthermore, The deletion of a changed line is possible which might be useful in a case where an identation fix really doesn't matter if a line is not needed. I consider this an expected but maybe dangerous behavior therefore this has to be discussed. Maybe the user should be warned in this case.

The diff-class should furthermore provide a method that returns the file in question with all changes applied. This could also be realised through making the class an iterator.

**Comment by sils1297** [comment]
Proposal looks good, in most cases where you are not sure if this should be a conflict I'm in favor of making it a conflict.

There's still one issue to solve:

Do we want to apply diffs to diffs or diffs after each other to files?

**Comment by smfn3321** [comment]
In this case diffs to diffs definitely

it is 
* more performant since changes that do not affect the same line can easily be merged from a number of diffs into one
* the only way to use the feature to merge to changes in the same line

If you mean the difflib.diffs() for changes, I think that while adding two diff-classes together both diffs for a change should be merged into a copy of the file and a diff for this file then has to be extracted and kept so that the new diff-class holds the expected diff for both changes

Concerning conflicts, the only think unclear to me is how to handle two diff-classes with one having a change for a line and one deleting it. I would prefer it to ignore the change in this case. Can you tell me a situation in which this would be nonobvious and destructable?

**Comment by sils1297** [comment]
We'll have to be careful with conflicts. When in doubt I'd really rather
ask the user instead of applying possibly unwanted changes.

2015-01-20 19:21 GMT+01:00 Fabian Neuschmidt <notifications@github.com>:

**Comment by sils1297** [comment]
depends on #168

**Comment by sils1297** [comment]
please take a look at #185. This would mean we're applying diffs to the file dict and if we have subsequent diffs we'll inspect the original and modified file to calculate offsets which in turn will allow us to apply diffs more precisely.

**Comment by sils1297** [comment]
ok, diffs are finished, they can be merged easily too. (__add__)

**Comment by sils1297** [comment]
diffs are there, PatchResult is on a #190. The result and diff part is done, conflict behaviour could be improved but thats left for later times. Outputter support isn't there yet but will coome with #168 . So I'll close this.
